packages/common/context/src/resource.ts

```diff
@@ -0,0 +1,131 @@
+//
+// Copyright 2024 DXOS.org
+//
+
+import '@dxos/util';
+import { Context } from './context';
+
+export interface Lifecycle {
+  open?(ctx?: Context): Promise<any> | any;
+  close?(): Promise<any> | any;
+}
+
+export enum LifecycleState {
+  CLOSED = 'CLOSED',
+  OPEN = 'OPEN',
+  ERROR = 'ERROR',
+}
+
+/**
+ * Base class for resources that need to be opened and closed.
+ */
+export class Resource implements Lifecycle {
+  #lifecycleState = LifecycleState.CLOSED;
+  #openPromise: Promise<void> | null = null;
+  #closePromise: Promise<void> | null = null;
+  #ctx: Context = this.#createContext();
```

Comment:
not sure I understand what this context is for. also I think parentContext should be created using createContext?

Reply:
I've added jsdoc comments for them